File: advent-of-code/2015/03/01/solution.zig
const std = @import("std");

fn countHouses(directions: []const u8) !u32 {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var x: i32 = 0;
    var y: i32 = 0;
    var visited = std.AutoHashMap([2]i32, void).init(allocator);
    defer visited.deinit();

    try visited.put([2]i32{ 0, 0 }, {});

    for (directions) |c| {
        switch (c) {
            '^' => y += 1,
            'v' => y -= 1,
            '>' => x += 1,
            '<' => x -= 1,
            else => {},
        }
        try visited.put([2]i32{ x, y }, {});
    }

    return visited.count();
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();

    // Test cases with assertions
    std.debug.assert(try countHouses(">") == 2);
    std.debug.assert(try countHouses("^>v<") == 4);
    std.debug.assert(try countHouses("^v^v^v^v^v") == 2);

    const input = try std.fs.cwd().readFileAlloc(gpa.allocator(), "../input.txt", 1024 * 1024);
    defer gpa.allocator().free(input);
    const directions = std.mem.trim(u8, input, &std.ascii.whitespace);

    const result = try countHouses(directions);
    std.debug.print("Zig: {d}\n", .{result});
}

File: advent-of-code/2015/03/01/solution.clj
(ns solution
  (:require [clojure.java.io :as io]
            [clojure.string :as str]))

(defn count-houses [directions]
  (let [moves {\^ [0 1] \v [0 -1] \> [1 0] \< [-1 0]}]
    (loop [x 0 y 0 visited #{[0 0]} dirs (seq directions)]
      (if (empty? dirs)
        (count visited)
        (let [[dx dy] (get moves (first dirs) [0 0])
              new-x (+ x dx)
              new-y (+ y dy)]
          (recur new-x new-y (conj visited [new-x new-y]) (rest dirs)))))))

;; Test cases with assertions
(assert (= 2 (count-houses ">")))
(assert (= 4 (count-houses "^>v<")))
(assert (= 2 (count-houses "^v^v^v^v^v")))

;; Read input and calculate result
(let [input-path (str (.getParent (io/file *file*)) "/../input.txt")
      directions (str/trim (slurp input-path))
      result (count-houses directions)]
  (println (str "Clojure: " result)))
